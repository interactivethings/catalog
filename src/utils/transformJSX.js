import React from 'react';
import {transform} from 'babel-standalone';
import requireModuleDefault from './requireModuleDefault';

const presets = ['es2015-loose', 'react', 'stage-2'];

// Babel plugin to return last top-level expression statement
const returnLastExpressionPlugin = ({types: t}) => ({
  visitor: {
    // We only care about top-level expressions
    Program(path) {
      // Find the last expression statement
      let lastExpr;
      for (let i = path.node.body.length - 1; i >= 0; i--) {
        if (t.isExpressionStatement(path.node.body[i])) {
          lastExpr = path.get(`body.${i}`);
          break;
        }
      }

      if (lastExpr) {
        // ... and turn it into a return statement
        lastExpr.replaceWith(
          t.returnStatement(
            lastExpr.node.expression
          )
        );
      }
    }
  }
});

let cached = {};
const cachedTransform = (jsx) => {
  if (cached[jsx]) {
    return cached[jsx];
  }
  const transformed = transform(jsx, {compact: true, presets, plugins: [returnLastExpressionPlugin]}).code;
  cached[jsx] = transformed;
  return transformed;
};

const missingTransformError = {
  error: 'Please include [babel-standalone](https://github.com/babel/babel-standalone) before Catalog.'
};

export default (jsx, imports) => {
  // Check for transform to provide a better error message
  try {
    transform;
  } catch (error) {
    return missingTransformError;
  }

  try {
    const importKeys = Object.keys(imports).filter((k) => imports[k]);
    const importModules = importKeys.map((k) => requireModuleDefault(imports[k]));
    const code = cachedTransform(jsx);
    const element = (new Function('React', ...importKeys, code))(React, ...importModules); // eslint-disable-line no-new-func
    return {code, element};
  } catch (error) {
    return {error};
  }
};
