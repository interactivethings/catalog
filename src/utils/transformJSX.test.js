import transformJSX from './transformJSX';

test('Transforms a simple element', () => {
  expect(transformJSX('<div>foo</div>', {}).code).toMatchSnapshot();
});

test('Transforms a nested element', () => {
  expect(transformJSX('<div><p>foo</p></div>', {}).code).toMatchSnapshot();
});

test('Transforms a component', () => {
  expect(transformJSX('<Foo bar>baz</Foo>', {Foo: () => null}).code).toMatchSnapshot();
});

test('Transforms a component which is defined inline', () => {
  expect(transformJSX(`const Foo = () => null;
<Foo bar>baz</Foo>`, {}).code).toMatchSnapshot();
});

test('Transforms an array of elements', () => {
  expect(transformJSX(`[
    <div key={1}>hey</div>,
    <div key={2}>ho</div>
  ]`, {}).code).toMatchSnapshot();
});

test('Returns the last expression', () => {
  expect(transformJSX(`
    React.createElement('div', {}, 'hey');
    React.createElement('div', {}, 'ho');
  `, {}).code).toMatchSnapshot();
});

test('Returns any expression', () => {
  expect(transformJSX(`
    const h = React.createElement;
    h('div', {}, 'hey');
  `, {}).code).toMatchSnapshot();
});

test('Returns error on syntax error', () => {
  expect(transformJSX('<div>', {}).error).toBeDefined();
});

test('Returns error when import is missing', () => {
  expect(transformJSX('<Foo />', {}).error).toBeDefined();
});
