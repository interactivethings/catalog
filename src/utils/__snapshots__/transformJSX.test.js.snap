// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Returns any expression 1`] = `"'use strict';var h=React.createElement;return h('div',{},'hey');"`;

exports[`Returns the last expression 1`] = `"'use strict';React.createElement('div',{},'hey');return React.createElement('div',{},'ho');"`;

exports[`Transforms a component 1`] = `"\\"use strict\\";return React.createElement(Foo,{bar:true},\\"baz\\");"`;

exports[`Transforms a component which is defined inline 1`] = `"\\"use strict\\";var Foo=function Foo(){return null;};return React.createElement(Foo,{bar:true},\\"baz\\");"`;

exports[`Transforms a nested element 1`] = `"\\"use strict\\";return React.createElement(\\"div\\",null,React.createElement(\\"p\\",null,\\"foo\\"));"`;

exports[`Transforms a simple element 1`] = `"\\"use strict\\";return React.createElement(\\"div\\",null,\\"foo\\");"`;

exports[`Transforms an array of elements 1`] = `"\\"use strict\\";return[React.createElement(\\"div\\",{key:1},\\"hey\\"),React.createElement(\\"div\\",{key:2},\\"ho\\")];"`;
